File: src/x/crazyjvm/factory/s/Factory.scala
package x.crazyjvm.factory.s

/**
 * Created by chenchao on 15/10/6.
 */
class Factory {

}

trait Animal
private class Dog extends Animal
private class Cat extends Animal

object Animal {
  def apply(kind: String) = kind match {
    case "dog" => new Dog()
    case "cat" => new Cat()
  }
}

object Factory extends App {
  Animal("dog")
}

File: src/x/crazyjvm/factory/j/Factory.java
package x.crazyjvm.factory.j;

/**
 * Created by chenchao on 15/10/6.
 */


class Dog implements Animal {}

class Cat implements Animal {}


class AnimalFactory {
    public static Animal createAnimal(String kind) {
        if ("cat".equals(kind)) return new Cat();
        if ("dog".equals(kind)) return new Dog();
        throw new IllegalArgumentException();
    }
}

public class Factory {
    public static void main(String[] args) {
        AnimalFactory.createAnimal("dog");
    }
}
